src/compiler/transformers/declarations.ts

```diff
@@ -725,45 +741,197 @@ export function transformDeclarations(context: TransformationContext) {
+                    typeNode = declaredType ? visitTypeNode(declaredType) : inferReturnTypeOfSignatureSignature(node);
+                    break;
+                case SyntaxKind.ExportAssignment:
+                    typeNode = typeFromExpression(node.expression) ?? inferExpressionType(node.expression);
+                    break;
+                default:
+                    Debug.assertNever(node, `Node needs to be an inferrable node, found ${Debug.formatSyntaxKind((node as Node).kind)}`);
+            }
+            errorNameNode = savedErrorNameNode;
+            errorFallbackNode = savedErrorFallbackNode;
+
+            if (!suppressNewDiagnosticContexts) {
+                getSymbolAccessibilityDiagnostic = oldDiag!;
+            }
+            return typeNode;
         }
-        return typeNode ?? factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);
-    }
 
+        function visitTypeNode(type: TypeNode) {
+            return visitNode(type, visitDeclarationSubtree, isTypeNode)!;
+        }
+
+        function addUndefinedInUnion(type: TypeNode) {
+            if (!strictNullChecks) return type;
+            if (isUnionTypeNode(type)) {
+                const hasUndefined = type.types.some(p => p.kind === SyntaxKind.UndefinedKeyword);
+                if (hasUndefined) return type;
+
+                return factory.createUnionTypeNode([
+                    ...type.types,
+                    factory.createKeywordTypeNode(SyntaxKind.UndefinedKeyword),
+                ]);
+            }
+            return factory.createUnionTypeNode([
+                type,
+                factory.createKeywordTypeNode(SyntaxKind.UndefinedKeyword),
+            ]);
+        }
+
+        function canAddUndefined(node: TypeNode): boolean {
+            if (!strictNullChecks) return true;
+            if (
+                isKeyword(node.kind)
+                || node.kind === SyntaxKind.LiteralType
+                || node.kind === SyntaxKind.FunctionType
+                || node.kind === SyntaxKind.ConstructorType
+                || node.kind === SyntaxKind.ArrayType
+                || node.kind === SyntaxKind.TupleType
+                || node.kind === SyntaxKind.TypeLiteral
+                || node.kind === SyntaxKind.TemplateLiteralType
+                || node.kind === SyntaxKind.ThisType
+            ) {
+                return true;
+            }
+            if (node.kind === SyntaxKind.ParenthesizedType) {
+                return canAddUndefined((node as ParenthesizedTypeNode).type);
+            }
+            if (node.kind === SyntaxKind.UnionType || node.kind === SyntaxKind.IntersectionType) {
+                return (node as UnionTypeNode | IntersectionTypeNode).types.every(canAddUndefined);
+            }
+            return false;
+        }
+        function typeFromExpression(node: Expression, requiresUndefined?: boolean): TypeNode | undefined {
+            switch (node.kind) {
+                case SyntaxKind.AsExpression:

```

Comment:
I think you're missing a `case` here

```suggestion
                case SyntaxKind.AsExpression:
                case SyntaxKind.TypeAssertion:
```

Reply:
Yup. Dropped it accidentally when moved code over.